src/main/java/io/numaproj/numaflow/function/FunctionService.java

```diff
@@ -154,59 +152,56 @@ public StreamObserver<Udfunction.Datum> reduceFn(StreamObserver<Udfunction.Datum
         IntervalWindow iw = new IntervalWindowImpl(startTime, endTime);
         Metadata md = new MetadataImpl(iw);
 
-        List<Future<Message[]>> results = new ArrayList<>();
+        CompletableFuture<Void> failureFuture = new CompletableFuture<>();
+
+        // create a shutdown actor that listens to exceptions.
+        ActorRef shutdownActorRef = actorSystem.
+                actorOf(ShutdownActor.props(responseObserver, failureFuture));
+
+        handleFailure(failureFuture);
+        /*
+            create a supervisor actor which assign the tasks to child actors.
+            we create a child actor for every key in a window.
+        */
+        ActorRef parentActorRef = actorSystem
+                .actorOf(ReduceSupervisorActor.props(groupBy, md, shutdownActorRef));
```

Comment:
I believe we should definitely consider a singleton Factory instance which can be pre-configured (similar to a builder with necessary user specific configuration) and then something that can expose a constructor to receive the arguments per key.

Reply:
can be in a separate PR.

Reply:
Okay